### ISSUE TITLE
Resetting some model parameters has no effect

### ISSUE THREAD

**Issue by rplzzz** [JGCRI/hector#260]
What's happening here is that if you reset a parameter like `PREINDUSTRIAL_CO2` and then call `reset(0)`, the new spinup gets run with the _old_ value of the preindustrial concentration.  This is because the preindustrial value is only referred to in `prepareToRun`.  An easy fix would be to rerun `prepareToRun` before we run the spinup in a reset, but before we can do that we need to know whether it's safe to do so.

**Question: Is it safe to rerun `prepareToRun` as part of a reset?**

_Update: It looks like almost all of the model parameters are referenced only in `prepareToRun`, so we're going to have to check that it's safe to run and make it safe if it's not.  Here's the checklist:

- [x] bc_component.cpp
- [x] carbon-cycle-solver.cpp
- [x] ch4_component.cpp
- [x] core.cpp
- [x] dummy_model_component.cpp
- [x] forcing_component.cpp
- [x] halocarbon_component.cpp
- [x] n2o_component.cpp
- [x] o3_component.cpp
- [x] oc_component.cpp
- [x] ocean_component.cpp
- [x] oh_component.cpp
- [x] onelineocean_component.cpp
- [x] simpleNbox.cpp
- [x] slr_component.cpp
- [x] so2_component.cpp
- [x] temperature_component.cpp

**Comment by rplzzz** [comment]
Issues

- [ ] Ocean box setup assumes that the boxes are pristine.  The `carbonHistory` member and connection variables may need to be reset before the method is run.

**Comment by bpbond** [comment]
>Is it safe to rerun prepareToRun as part of a reset?

Looks like you're working through the components and checking them, but yeah, I'd assume the answer is currently "no". Our original Hector design never anticipated multiple calls to `prepareToRun()`.

**Comment by rplzzz** [comment]
What has happened here is that we have three initialization steps (class constructor, `init`, and `prepareToRun` with no clear guidelines for what kinds of things are supposed to go in each step.  I think the breakdown should be something like

* **constructor**: minimal setup to ensure that all member variables are initialized to a known state
* **init**: establish connections between components and the core (including subcomponents, so the ocean boxes I mentioned above should go here).  This should only be run once.
* **prepareToRun**: initialize state variables (if applicable), cache connections to other components (if applicable).  This should be safe to run multiple times (e.g., after a reset).

One obvious question is, do we really need `init` to be separate from the constructor?  All of the components are constructed in `Core::init` anyway, and there doesn't seem to be anything happening in between constructing the components and calling the components' `init` methods, so we could simplify the setup process by doing the work in `init` from the constructors.

Another wrinkle is that we don't read the input file until after `Core::init` is run.  It's not clear to me whether it _has_ to be this way.  The handling of the parsing is hard to follow, as it passes through several classes and some plain C code.  It's not so easy to figure out whether it needs the component wiring to be set up, or if it's just done where it is because someone decided to do it that way.  Assuming that the parser has to be run after the wiring is done, this forces us to put any setup that depends on the input in `prepareToRun`.  Mostly that's what we want, but it could be a problem if there is any structure setup that is being controlled by the input.

**Comment by bpbond** [comment]
>with no clear guidelines for what kinds of things are supposed to go in each step

Well, there _were_ quite clear guidelines in Hector's design:
* The constructor is minimal initialization
* `init` is to establish connections and is run only once. At this point it's guaranteed that all other components are instantiated
* `prepareToRun` is for any work necessary to be ready to start, and is run once once. At this point it's guaranteed that all input is finished and available, and component dependencies are established

You raise a good point about the need for separate constructor and `init` functions; we may be able to collapse their code into `init`. 

As I noted above, the original design didn't anticipate calling `prepareToRun` more than once. That's a spec chance necessitated by an evolving model.
